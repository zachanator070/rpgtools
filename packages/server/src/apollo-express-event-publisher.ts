import { EventPublisher } from "./types";
import { PubSub } from "apollo-server-express";
import { injectable } from "inversify";

@injectable()
export class ApolloExpressEventPublisher implements EventPublisher {
	pubsub = new PubSub();

	asyncIterator(events: string[]): AsyncIterator<any> {
		return this.pubsub.asyncIterator(events);
	}

	publish(event: string, payload: any): Promise<void> {
		return this.pubsub.publish(event, payload);
	}
}
