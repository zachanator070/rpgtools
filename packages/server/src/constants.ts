export type CookieConstants = {
	string: string;
	ms: number;
};

export const ACCESS_TOKEN: string = "accessToken";
export const REFRESH_TOKEN: string = "refreshToken";

export const ACCESS_TOKEN_MAX_AGE: CookieConstants = { string: "1m", ms: 1000 * 60 };
export const REFRESH_TOKEN_MAX_AGE: CookieConstants = { string: "1d", ms: 1000 * 60 * 60 * 24 };
