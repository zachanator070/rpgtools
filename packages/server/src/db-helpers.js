import { PermissionAssignment } from "./models/permission-assignement";
import { Role } from "./models/role";
import { User } from "./models/user";
import mongodb, { GridFSBucket } from "mongodb";
import mongoose from "mongoose";

export const cleanUpPermissions = async (subjectId) => {
  const assignments = await PermissionAssignment.find({ subject: subjectId });
  for (let assignment of assignments) {
    const roles = await Role.find({ permissions: assignment._id });
    for (let role of roles) {
      role.permissions = role.permissions.filter((permission) => {
        return !permission._id.equals(assignment._id);
      });
      await role.save();
    }
    const users = await User.find({ permissions: assignment._id });
    for (let user of users) {
      user.permissions = user.permissions.filter((permission) => {
        return !permission._id.equals(assignment._id);
      });
      await user.save();
    }
    await PermissionAssignment.deleteOne({ _id: assignment._id });
  }
};

export const createGfsFile = async (filename, readStream) => {
  return new Promise((resolve, reject) => {
    const gfs = new GridFSBucket(mongoose.connection.db);
    const writeStream = gfs.openUploadStream(filename);

    writeStream.on("finish", (file) => {
      resolve(file._id);
    });

    writeStream.on("error", (err) => {
      reject(err);
      throw err;
    });

    readStream.pipe(writeStream);
  });
};

export const deleteGfsFile = async (fileId) => {
  return new Promise((resolve, reject) => {
    const gfs = new GridFSBucket(mongoose.connection.db);

    gfs.delete(new mongoose.Types.ObjectId(fileId), (error) => {
      resolve();
    });
  });
};
export const getGfsFileFromFileId = async (fileId) => {
  return new Promise((resolve, reject) => {
    const gfs = new GridFSBucket(mongoose.connection.db);
    const id = new mongodb.ObjectID(fileId);
    gfs.find({ _id: id }).next((err, file) => {
      if (err) {
        reject(err);
        return;
      }
      if (file === null) {
        reject(`file not found with id ${fileId}`);
        return;
      }
      resolve(file);
    });
  });
};
export const getReadStreamFromFile = (file) => {
  const gfs = new GridFSBucket(mongoose.connection.db);
  return gfs.openDownloadStream(file._id);
};
