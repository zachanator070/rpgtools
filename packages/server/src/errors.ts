export class EntityNotFoundError extends Error {
	constructor(id: string, type: string) {
		super(`${type} with id ${id} not found`);
	}
}

export class ReadPermissionDeniedError extends Error {
	constructor(id: string, type: string) {
		super(`You do not have read permission for ${type} with id ${id}`);
	}
}

export class WritePermissionDeniedError extends Error {
	constructor(id: string, type: string) {
		super(`You do not have write permission for ${type} with id ${id}`);
	}
}

export class AdminPermissionDeniedError extends Error {
	constructor(id: string, type: string) {
		super(`You do not have admin permission for ${type} with id ${id}`);
	}
}

export class TypeNotSupportedError extends Error {
	constructor(type: string) {
		super(`This function does not support the type ${type}`);
	}
}
