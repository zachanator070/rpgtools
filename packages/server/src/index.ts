import {container} from "./inversify.config";
import {INJECTABLE_TYPES} from "./injectable-types";
import {ApiServer} from "./types";

const server = container.get<ApiServer>(INJECTABLE_TYPES.ApiServer);

(async () => {
	await server.start();
})();
