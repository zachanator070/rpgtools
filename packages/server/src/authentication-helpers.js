import jwt from "jsonwebtoken";
import { v4 as uuidv4 } from "uuid";
import { User } from "./models/user";
import { ANON_USERNAME } from "@rpgtools/common/src/permission-constants";

export const authenticated = (next) => (root, args, context, info) => {
  if (!context.currentUser || context.currentUser.username === ANON_USERNAME) {
    throw new Error(`Authentication is required for this action`);
  }

  return next(root, args, context, info);
};

export const createTokens = async (user) => {
  const accessToken = jwt.sign(
    { userId: user._id },
    process.env["ACCESS_TOKEN_SECRET"],
    { expiresIn: ACCESS_TOKEN_MAX_AGE.string }
  );
  let version = uuidv4();
  const refreshToken = jwt.sign(
    { version: version, userId: user._id },
    process.env["REFRESH_TOKEN_SECRET"],
    { expiresIn: REFRESH_TOKEN_MAX_AGE.string }
  );
  user.tokenVersion = version;
  await user.save();
  return { accessToken, refreshToken };
};

export const getCurrentUser = async (userId) => {
  return User.findById(userId)
    .populate({ path: "currentWorld", populate: { path: "wikiPage" } })
    .populate({ path: "permissions", populate: "subject" })
    .populate({ path: "roles", populate: "permissions" });
};

export const createSessionContext = async ({ req, res, connection }) => {
  let currentUser = null;
  if (connection) {
    return connection.context;
  }

  const refreshToken = req.cookies["refreshToken"];
  const accessToken = req.cookies["accessToken"];

  // if access token wasn't given but refresh token was
  if (accessToken || refreshToken) {
    try {
      let data = jwt.verify(accessToken, process.env["ACCESS_TOKEN_SECRET"], {
        maxAge: ACCESS_TOKEN_MAX_AGE.string,
      });
      currentUser = await getCurrentUser(data.userId);
    } catch (e) {
      // accessToken is expired
      console.log(`Access token expired because ${e.message}`);
      res.clearCookie("accessToken");
      try {
        let data = jwt.verify(
          refreshToken,
          process.env["REFRESH_TOKEN_SECRET"],
          { maxAge: REFRESH_TOKEN_MAX_AGE.string }
        );
        currentUser = await getCurrentUser(data.userId);
        // if refreshToken is still valid issue new access token and refresh token
        if (currentUser && currentUser.tokenVersion === data.version) {
          console.log(
            "Access token and refresh token renewed b/c refresh token still valid"
          );
          let tokens = await createTokens(currentUser);
          res.cookie("accessToken", tokens.accessToken, {
            maxAge: ACCESS_TOKEN_MAX_AGE.ms,
            secure: false,
          });
          res.cookie("refreshToken", tokens.refreshToken, {
            maxAge: REFRESH_TOKEN_MAX_AGE.ms,
            secure: false,
          });
        } else {
          // refreshToken was invalidated
          console.log(
            `Refresh token expired because version did not match ${currentUser.tokenVersion} !== ${data.version}`
          );
          currentUser = null;
          res.clearCookie("refreshToken");
        }
      } catch (e) {
        // refreshToken is expired
        console.log(`Refresh token expired because ${e.message}`);
        throw new Error("Refresh token expired");
      }
    }
  }

  if (!currentUser) {
    currentUser = new User({ username: ANON_USERNAME });
  }

  await currentUser.recalculateAllPermissions();

  return {
    res,
    currentUser,
  };
};

// export const ACCESS_TOKEN_MAX_AGE = {string: '15min', ms: 1000 * 60 * 15};
export const ACCESS_TOKEN_MAX_AGE = { string: "1m", ms: 1000 * 60 };
export const REFRESH_TOKEN_MAX_AGE = { string: "1d", ms: 1000 * 60 * 60 * 24 };
