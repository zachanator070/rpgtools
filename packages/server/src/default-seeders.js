import { Role } from "./models/role";
import { PermissionAssignment } from "./models/permission-assignement";
import { ALL_USERS } from "@rpgtools/common/src/role-constants";
import { WORLD_CREATE } from "@rpgtools/common/src/permission-constants";
import { ServerConfig } from "./models/server-config";
import { SERVER_CONFIG } from "@rpgtools/common/src/type-constants";
import { v4 as uuidv4 } from "uuid";

export const seedDefaultRoles = async () => {
  let allUsersRole = await Role.find({ name: ALL_USERS });
  if (allUsersRole.length === 0) {
    const server = await ServerConfig.findOne();
    if (!server) {
      throw new Error("Server needs to exist!");
    }
    allUsersRole = await Role.create({ name: ALL_USERS });
    let createWorldPermission = await PermissionAssignment.findOne({
      permission: WORLD_CREATE,
      subject: server._id,
      subjectType: SERVER_CONFIG,
    });
    if (!createWorldPermission) {
      createWorldPermission = await PermissionAssignment.create({
        permission: WORLD_CREATE,
        subject: server._id,
        subjectType: SERVER_CONFIG,
      });
    }
    allUsersRole.permissions.push(createWorldPermission._id);
    await allUsersRole.save();
    console.log(`Created default role "${ALL_USERS}"`);
  }
};

export const seedServer = async () => {
  let server = await ServerConfig.findOne();
  if (!server) {
    server = await ServerConfig.create({
      unlockCode: uuidv4(),
      version: "1.0",
    });
    console.log(`Created server config ${server._id}`);
    console.log(`====================================================`);
    console.log(`Server unlock code: ${server.unlockCode}`);
    console.log(`====================================================`);
  }
};

export const defaultSeeders = async () => {
  await seedServer();
  await seedDefaultRoles();
};
