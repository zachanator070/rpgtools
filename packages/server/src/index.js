import express from "express";
import bodyParser from "body-parser";
import gqlServer from "./gql-server";
import cookieParser from "cookie-parser";
import morgan from "morgan";
import mongoose from "mongoose";
import { defaultSeeders } from "./default-seeders";
import path from "path";
import { ImageRouter } from "./routers/image-router";
import { checkConfig, serverNeedsSetup } from "./server-needs-setup";
import crypto from "crypto";
import * as http from "http";
import { ServerConfig } from "./models/server-config";
import { ModelRouter } from "./routers/model-router";
import ExportRouter from "./routers/export-router";
import { graphqlUploadExpress } from "graphql-upload";

const mongodb_host = process.env.MONGODB_HOST || "mongodb";
const mongodb_db_name = process.env.MONGODB_DB_NAME || "rpgtools";

mongoose
  .connect(`mongodb://${mongodb_host}/${mongodb_db_name}`, {
    useNewUrlParser: true,
  })
  .then(async () => {
    console.log(
      `Connected to mongodb at mongodb://${mongodb_host}/${mongodb_db_name}`
    );
    await defaultSeeders();
    await createServer();
  });

mongoose.set("useNewUrlParser", true);
mongoose.set("useFindAndModify", false);
mongoose.set("useCreateIndex", true);
mongoose.set("useUnifiedTopology", true);

const createServer = async () => {
  const server = express();

  server.use(bodyParser.json());
  server.use(morgan("tiny"));

  server.use(cookieParser());

  await checkConfig();
  const serverConfig = await ServerConfig.findOne();
  if (serverNeedsSetup()) {
    console.warn(
      `Server needs configuration! Use unlock code ${serverConfig.unlockCode} to unlock`
    );
  }

  server.get("*.js", function (req, res, next) {
    req.url = req.url + ".gz";
    res.set("Content-Encoding", "gzip");
    res.set("Content-Type", "text/javascript");
    next();
  });
  server.get("*.css", function (req, res, next) {
    req.url = req.url + ".gz";
    res.set("Content-Encoding", "gzip");
    res.set("Content-Type", "text/css");
    next();
  });
  server.get("*favicon.ico", function (req, res, next) {
    req.url = "/favicon.ico";
    next();
  });

  server.use("/images", ImageRouter);
  server.use("/models", ModelRouter);
  server.use("/export", ExportRouter);

  server.get("*", (req, res, next) => {
    const needsSetup = serverNeedsSetup();
    if (needsSetup) {
      if (
        [
          "/api",
          "/ui/setup",
          "/app.bundle.js",
          "/app.css",
          "/app.bundle.js.gz",
          "/app.css.gz",
          "favicon.ico",
        ].includes(req.url)
      ) {
        return next();
      }
      return res.redirect(302, "/ui/setup");
    } else {
      return next();
    }
  });

  server.get("/ui*", (req, res) => {
    return res.sendFile(path.resolve(__dirname, "../../dist", "index.html"));
  });

  server.use(express.static("dist"));
  server.use(express.static("src/static-assets"));
  server.use(graphqlUploadExpress());

  gqlServer.applyMiddleware({ app: server, path: "/api" });

  if (!process.env.ACCESS_TOKEN_SECRET) {
    console.warn(
      "environment variable ACCESS_TOKEN_SECRET is not set, restarting server will log out all users"
    );
    process.env.ACCESS_TOKEN_SECRET = crypto.randomBytes(2048);
  }

  if (!process.env.REFRESH_TOKEN_SECRET) {
    console.log(
      "environment variable REFRESH_TOKEN_SECRET is not set, restarting server will log out all users"
    );
    process.env.REFRESH_TOKEN_SECRET = crypto.randomBytes(2048);
  }

  const port = process.env.SERVER_PORT || 3000;

  const httpServer = http.createServer(server);
  gqlServer.installSubscriptionHandlers(httpServer);

  httpServer.listen(port, () => {
    console.log(
      `The server is running and listening at http://localhost:${port}`
    );
  });
};
