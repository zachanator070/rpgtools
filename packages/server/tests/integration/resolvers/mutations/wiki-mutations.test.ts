import { Readable } from "stream";
import fs from "fs";
import { ARTICLE } from "../../../../../common/src/type-constants";
import { CREATE_WIKI } from "../../../../../frontend/src/hooks/wiki/useCreateWiki";
import { DELETE_WIKI } from "../../../../../frontend/src/hooks/wiki/useDeleteWiki";
import { UPDATE_PLACE } from "../../../../../frontend/src/hooks/wiki/useUpdatePlace";
import { UPDATE_WIKI } from "../../../../../frontend/src/hooks/wiki/useUpdateWiki";
import {CREATE_IMAGE} from "../../../../../common/src/mutations";
import {defaultTestingContextFactory} from "../../DefaultTestingContextFactory";

process.env.TEST_SUITE = "wiki-mutations-test";

describe("wiki page mutations", () => {
	let {
		server,
		mockSessionContextFactory,
		otherUser,
		otherUserSecurityContext,
		world,
		testRole,
		currentUser,
		testerSecurityContext,
		newFolder,
		otherPage,
		...testingContext
	} = defaultTestingContextFactory();

	describe("with world and not logged in", () => {
		beforeEach(async () => {
			({
				mockSessionContextFactory,
				otherUser,
				otherUserSecurityContext,
				world,
				testRole,
				currentUser,
				testerSecurityContext,
				newFolder,
				otherPage,
			} = await testingContext.reset());
			mockSessionContextFactory.resetCurrentUser();
		});

		test("create wiki no permission", async () => {
			const result = await server.executeGraphQLQuery({
				query: CREATE_WIKI,
				variables: { name: "new wiki", folderId: world.rootFolder.toString() },
			});
			expect(result).toMatchSnapshot();
		});

		test("update wiki no permission", async () => {
			const result = await server.executeGraphQLQuery({
				query: UPDATE_WIKI,
				variables: { wikiId: world.wikiPage, name: "new name" },
			});
			expect(result).toMatchSnapshot();
		});

		test("update cover image no permission", async () => {
			const result = await server.executeGraphQLQuery({
				query: UPDATE_WIKI,
				variables: {
					wikiId: world.wikiPage,
					coverImageId: "12345",
				},
			});
			expect(result).toMatchSnapshot();
		});

		test("delete wiki no permission", async () => {
			const result = await server.executeGraphQLQuery({
				query: DELETE_WIKI,
				variables: { wikiId: world.wikiPage},
			});
			expect(result).toMatchSnapshot();
		});

		test("update place no permission", async () => {
			const result = await server.executeGraphQLQuery({
				query: UPDATE_PLACE,
				variables: { placeId: world.wikiPage, mapImageId: null },
			});
			expect(result).toMatchSnapshot();
		});

		describe("with authenticated user", () => {
			beforeEach(async () => {
				mockSessionContextFactory.setCurrentUser(currentUser);
			});

			test("create wiki", async () => {
				const result = await server.executeGraphQLQuery({
					query: CREATE_WIKI,
					variables: {
						name: "new wiki",
						folderId: world.rootFolder.toString(),
					},
				});
				expect(result).toMatchSnapshot({
					data: {
						createWiki: {
							_id: expect.any(String),
							folders: expect.arrayContaining([
								expect.objectContaining({
									_id: expect.any(String),
									pages: expect.arrayContaining([
										expect.objectContaining({
											_id: expect.any(String),
										}),
									]),
								}),
							]),
						},
					},
				});
			});

			test("update wiki just name", async () => {
				const result = await server.executeGraphQLQuery({
					query: UPDATE_WIKI,
					variables: {
						wikiId: world.wikiPage,
						name: "new name",
					},
				});
				expect(result).toMatchSnapshot({
					data: {
						updateWiki: {
							_id: expect.any(String),
							world: {
								_id: expect.any(String),
							},
						},
					},
				});
			});

			test("update wiki just content", async () => {
				const content = {
					createReadStream: () => Readable.from("wiki content".repeat(1024)),
				};
				const result = await server.executeGraphQLQuery({
					query: UPDATE_WIKI,
					variables: {
						wikiId: world.wikiPage,
						content: content,
					},
				});
				expect(result).toMatchSnapshot({
					data: {
						updateWiki: {
							_id: expect.any(String),
							world: {
								_id: expect.any(String),
							},
							content: expect.any(String),
						},
					},
				});
			});

			test("update cover image", async () => {
				let testFile = {
					filename: "server/tests/integration/resolvers/mutations/testmap.png",
					createReadStream: () =>
						fs.createReadStream("server/tests/integration/resolvers/mutations/testmap.png"),
				};
				const imageResult = await server.executeGraphQLQuery({
					query: CREATE_IMAGE,
					variables: {
						file: testFile,
						worldId: world._id.toString(),
						chunkify: true,
					},
				});
				const result = await server.executeGraphQLQuery({
					query: UPDATE_WIKI,
					variables: {
						wikiId: world.wikiPage,
						coverImageId: imageResult.data.createImage._id,
					},
				});
				expect(result).toMatchSnapshot({
					data: {
						updateWiki: {
							_id: expect.any(String),
							world: {
								_id: expect.any(String),
							},
							coverImage: {
								_id: expect.any(String),
								chunks: expect.arrayContaining([
									expect.objectContaining({
										_id: expect.any(String),
										fileId: expect.any(String),
									}),
								]),
								icon: {
									_id: expect.any(String),
									chunks: expect.arrayContaining([
										expect.objectContaining({
											_id: expect.any(String),
											fileId: expect.any(String),
										}),
									]),
								},
							},
						},
					},
				});
			});

			test("update type", async () => {
				const result = await server.executeGraphQLQuery({
					query: UPDATE_WIKI,
					variables: { wikiId: world.wikiPage, type: ARTICLE },
				});
				expect(result).toMatchSnapshot({
					data: {
						updateWiki: {
							_id: expect.any(String),
							world: {
								_id: expect.any(String),
							},
						},
					},
				});
			});

			test("delete wiki root wiki", async () => {
				const result = await server.executeGraphQLQuery({
					query: DELETE_WIKI,
					variables: { wikiId: world.wikiPage },
				});
				expect(result).toMatchSnapshot();
			});

			test("delete wiki", async () => {
				const result = await server.executeGraphQLQuery({
					query: DELETE_WIKI,
					variables: { wikiId: otherPage._id },
				});
				expect(result).toMatchSnapshot({
					data: {
						deleteWiki: {
							_id: expect.any(String),
							folders: expect.arrayContaining([
								expect.objectContaining({
									_id: expect.any(String),
									children: expect.arrayContaining([
										expect.objectContaining({
											_id: expect.any(String),
										}),
									]),
									pages: [],
								}),
								expect.objectContaining({
									_id: expect.any(String),
									children: [],
									pages: expect.arrayContaining([
										expect.objectContaining({
											_id: expect.any(String),
										}),
									]),
								}),
							]),
						},
					},
				});
			});

			test("update place", async () => {
				let testFile = {
					filename: "server/tests/integration/resolvers/mutations/testmap.png",
					createReadStream: () =>
						fs.createReadStream("server/tests/integration/resolvers/mutations/testmap.png"),
				};
				const imageResult = await server.executeGraphQLQuery({
					query: CREATE_IMAGE,
					variables: {
						file: testFile,
						worldId: world._id.toString(),
						chunkify: true,
					},
				});
				const result = await server.executeGraphQLQuery({
					query: UPDATE_PLACE,
					variables: {
						placeId: world.wikiPage,
						mapImageId: imageResult.data.createImage._id,
					},
				});
				expect(result).toMatchSnapshot({
					data: {
						updatePlace: {
							_id: expect.any(String),
							mapImage: {
								_id: expect.any(String),
								chunks: expect.arrayContaining([
									expect.objectContaining({
										_id: expect.any(String),
										fileId: expect.any(String),
									}),
								]),
								icon: {
									_id: expect.any(String),
									chunks: expect.arrayContaining([
										expect.objectContaining({
											_id: expect.any(String),
											fileId: expect.any(String),
										}),
									]),
								},
							},
						},
					},
				});
			});
		});
	});
});
