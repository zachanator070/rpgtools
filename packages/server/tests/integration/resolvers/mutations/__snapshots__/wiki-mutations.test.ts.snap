// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wiki page mutations with world and not logged in create wiki no permission 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission to write to the folder 6118afcd8ba4a42d70993651],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`wiki page mutations with world and not logged in delete wiki no permission 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission to write to this page],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`wiki page mutations with world and not logged in update cover image no permission 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Cannot read property 'createReadStream' of undefined],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`wiki page mutations with world and not logged in update place no permission 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission to write to this page],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`wiki page mutations with world and not logged in update wiki no permission 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Cannot read property 'createReadStream' of undefined],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`wiki page mutations with world and not logged in with authenticated user delete wiki root wiki 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You cannot delete the main page of a world],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
